Use FastCat in PT Concat implementation

zoranzhao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106518

Differential Revision: [D48081898](https://our.internmc.facebook.com/intern/diff/D48081898)

zoranzhao committed:
Use FastCat in PT Concat implementation

[ghstack-poisoned]

zoranzhao committed:
Update on "Use FastCat in PT Concat implementation"




[ghstack-poisoned]

zoranzhao committed:
Update on "Use FastCat in PT Concat implementation"




[ghstack-poisoned]

houseroad reviewed (commented):
Fix the lint?

zoranzhao committed:
Update on "Use FastCat in PT Concat implementation"




[ghstack-poisoned]

houseroad reviewed (commented):
Since this change is needed by internal use case first, let's create a diff then export to github? That will be faster to land in fbcode. Otherwise, we may need to wait one day or more to get it landed internally.

zoranzhao commented:
@zoranzhao has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48081898).

zoranzhao commented:
> Since this change is needed by internal use case first, let's create a diff then export to github? That will be faster to land in fbcode. Otherwise, we may need to wait one day or more to get it landed internally.

@houseroad I have created a diff in fbcode first [here](https://www.internalfb.com/diff/D48082741)

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -564,10 +616,24 @@ TORCH_IMPL_FUNC(cat_out_cpu)
   bool use_serial_kernel = result.numel() < at::internal::GRAIN_SIZE || at::get_num_threads() == 1;
   ScalarType dtype = materialized[valid].get().scalar_type();
   bool serial_dtype = at::isFloatingType(dtype);
+
```
This seems to get applied even when not single thread?

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -549,6 +549,58 @@ std::vector<Tensor> broadcast_tensors(TensorList tensors) {
+    if (cat_should_skip_tensor(input)) {
+      continue;
+    }
+    TORCH_CHECK(outBytes >= totalBytes);
```
This error check... seems kind of pointless? You are checking that the base of the range you are going to copy into is in bounds... but that's not what you really care about, you care about the end no?

ezyang reviewed (changes requested):
You need perf numbers. You are applying this fast path unconditionally even when the input is large enough to admit multithreading, need to show that it is still profitable in this case.

zoranzhao committed:
Update on "Use FastCat in PT Concat implementation"


Differential Revision: [D48081898](https://our.internmc.facebook.com/intern/diff/D48081898)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -564,11 +616,25 @@ TORCH_IMPL_FUNC(cat_out_cpu)
+  // fast path for single thread when both inputs and result are contiguous and
+  // not empty, and concat dim is 0 or 1
+  if (use_serial_kernel && all_contiguous && all_same_dtype &&
+      (MemoryFormat::Contiguous == memory_format) && (serial_dtype_int || serial_dtype_float)) {
```
Actually, why do you need to know the dtype at all? lol

houseroad replied:
Just to double check the complex types, their storage are also contiguous, so safe to use memcpy to copy over?

ezyang commented:
Switching to single threaded only satisfies one major question I had. The rest is just the benchmarks.

houseroad commented:
Good point, benchmark data is necessary, we can have small tensor benchmark + real case data. E2E QPS improvement is also useful here. Maybe we can attach that in the internal diff test plan.
